File: services/verifier/internal/verify/crypto.go
package verify

import (
	"crypto/ed25519"
	"math/big"

	"filippo.io/edwards25519"
)

// Ed25519 canonical-encoding and prime-order checks of event-types.md
// (ET-4a/ET-4b/ET-4c), run on the raw decoded bytes BEFORE the verification
// primitive (ET-5). Only ET-4c uses the edwards25519 curve library; every
// other check here is stdlib arithmetic, and ET-5 uses crypto/ed25519.

// L is the order of the Ed25519 prime-order subgroup:
// 2^252 + 27742317777372353535851937790883648493.
var edL = func() *big.Int {
	l := new(big.Int).Lsh(big.NewInt(1), 252)
	c, _ := new(big.Int).SetString("27742317777372353535851937790883648493", 10)
	return l.Add(l, c)
}()

// P is the field prime 2^255 - 19.
var edP = func() *big.Int {
	p := new(big.Int).Lsh(big.NewInt(1), 255)
	return p.Sub(p, big.NewInt(19))
}()

// scalarLminus1 is (L-1) as a canonical Ed25519 scalar; multiplying a point by
// it and adding the point back yields [L]A (see checkKeyPrimeOrder).
var scalarLminus1 = func() *edwards25519.Scalar {
	lm1 := new(big.Int).Sub(edL, big.NewInt(1))
	le := bigToLE32(lm1)
	s, err := edwards25519.NewScalar().SetCanonicalBytes(le)
	if err != nil {
		panic("verify: L-1 is not a canonical scalar: " + err.Error())
	}
	return s
}()

// bigToLE32 renders x as exactly 32 octets, little-endian.
func bigToLE32(x *big.Int) []byte {
	be := x.Bytes()
	le := make([]byte, 32)
	for i := 0; i < len(be); i++ {
		le[i] = be[len(be)-1-i]
	}
	return le
}

// leToBig interprets b as an unsigned little-endian integer.
func leToBig(b []byte) *big.Int {
	be := make([]byte, len(b))
	for i := 0; i < len(b); i++ {
		be[i] = b[len(b)-1-i]
	}
	return new(big.Int).SetBytes(be)
}

// checkSigCanonical implements ET-4a on the 64 decoded octets of sig:
// (i) S = sig[32:64] as LE integer is strictly < L; and
// (ii) R = sig[0:32] with bit 255 cleared, as LE integer, is strictly < p.
func checkSigCanonical(sig []byte) bool {
	if len(sig) != 64 {
		return false
	}
	s := leToBig(sig[32:64])
	if s.Cmp(edL) >= 0 {
		return false
	}
	r := make([]byte, 32)
	copy(r, sig[0:32])
	r[31] &= 0x7f // clear encoded sign bit (bit 255)
	if leToBig(r).Cmp(edP) >= 0 {
		return false
	}
	return true
}

// checkKeyCanonical implements ET-4b: the 32 decoded octets of A, with bit 255
// cleared, interpreted as an unsigned LE integer, are strictly < p.
func checkKeyCanonical(a []byte) bool {
	if len(a) != 32 {
		return false
	}
	y := make([]byte, 32)
	copy(y, a)
	y[31] &= 0x7f
	return leToBig(y).Cmp(edP) < 0
}

// checkKeyPrimeOrder implements ET-4c on an already-ET-4b-canonical key: A must
// lie in the prime-order subgroup, i.e. [L]A == identity AND A != identity.
//
// A Scalar in edwards25519 is reduced modulo L, so [L] cannot be applied
// directly. Instead [L]A is computed as [L-1]A + A, both representable.
func checkKeyPrimeOrder(a []byte) bool {
	A, err := new(edwards25519.Point).SetBytes(a)
	if err != nil {
		return false // not a decodable curve point
	}
	identity := edwards25519.NewIdentityPoint()
	if A.Equal(identity) == 1 {
		return false // A == O fails the non-identity clause (load-bearing, fixture 081)
	}
	la := new(edwards25519.Point).ScalarMult(scalarLminus1, A) // [L-1]A
	la.Add(la, A)                                              // [L]A
	return la.Equal(identity) == 1
}

// verifyEd25519 checks a signature under the given public key over msg, after
// the canonical (ET-4a/ET-4b) and prime-order (ET-4c) gates have passed. It
// returns false unless every gate and the ET-5 primitive succeed.
func verifyEd25519(pub, sig, msg []byte) bool {
	if len(pub) != 32 || len(sig) != 64 {
		return false
	}
	if !checkSigCanonical(sig) { // ET-4a
		return false
	}
	if !checkKeyCanonical(pub) { // ET-4b
		return false
	}
	if !checkKeyPrimeOrder(pub) { // ET-4c
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(pub), msg, sig) // ET-5
}

File: tools/verifier-ts/src/crypto.ts
// Ed25519 canonical-encoding, prime-order, and verification primitives.
//
// event-types.md ET-4a/ET-4b/ET-4c and ET-5. The canonical-encoding checks run
// on the RAW decoded bytes BEFORE the verification primitive is ever called, so
// the RFC-8032 underdetermination of non-canonical inputs is unreachable (the
// point of ET-4a/ET-4b). ET-4c (prime-order subgroup) needs curve scalar
// multiplication that node:crypto lacks, so it — and ONLY it — uses the one
// audited third-party curve library, @noble/curves. Everything else, including
// the ET-5 verify primitive, is node:crypto.

import { createPublicKey, verify as nodeVerify } from "node:crypto";
import { ed25519 } from "@noble/curves/ed25519.js";

const P = (1n << 255n) - 19n; // field prime 2^255 - 19
// order of the Ed25519 prime-order subgroup
const L = (1n << 252n) + 27742317777372353535851937790883648493n;

// Interpret octets as an unsigned little-endian integer.
function leToBigInt(bytes: Buffer): bigint {
  let acc = 0n;
  for (let i = bytes.length - 1; i >= 0; i--) {
    acc = (acc << 8n) | BigInt(bytes[i] as number);
  }
  return acc;
}

/**
 * ET-4b: canonical verification-key encoding. The 32 decoded octets of A, with
 * bit 255 (top bit of octet 31) cleared, interpreted little-endian, MUST be < p.
 * Runs on the raw key octets before any verification.
 */
export function isCanonicalKeyEncoding(raw32: Buffer): boolean {
  if (raw32.length !== 32) return false;
  const masked = Buffer.from(raw32);
  masked[31] = (masked[31] as number) & 0x7f;
  return leToBigInt(masked) < P;
}

/**
 * ET-4a: canonical signature encoding. Given the 64 octets R||S:
 *   (i)  S (little-endian) strictly < L
 *   (ii) R with bit 255 cleared (little-endian) strictly < p
 * Both must hold on the raw decoded bytes.
 */
export function isCanonicalSigEncoding(raw64: Buffer): boolean {
  if (raw64.length !== 64) return false;
  const R = raw64.subarray(0, 32);
  const S = raw64.subarray(32, 64);
  const sOk = leToBigInt(S) < L;
  const maskedR = Buffer.from(R);
  maskedR[31] = (maskedR[31] as number) & 0x7f;
  const rOk = leToBigInt(maskedR) < P;
  return sOk && rOk;
}

/**
 * ET-4c: prime-order verification key. On the ALREADY-canonical key (ET-4b has
 * passed), the decoded point A MUST satisfy [L]A == 𝒪 AND A != 𝒪 — i.e. it lies
 * in the prime-order subgroup and is not the identity. The `A != 𝒪` clause is
 * load-bearing: @noble's isTorsionFree() returns true for the identity, which
 * ET-4c must reject (fixture 081). Rejects all small-order AND mixed-order keys.
 */
export function isPrimeOrderKey(raw32: Buffer): boolean {
  try {
    const point = ed25519.Point.fromBytes(Uint8Array.from(raw32));
    return point.isTorsionFree() && !point.is0();
  } catch {
    // A non-decodable point cannot be in the subgroup.
    return false;
  }
}

// SPKI DER prefix for a raw 32-byte Ed25519 public key.
const SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

/**
 * ET-5: Ed25519 verification of `sig` over `message` under raw 32-byte key `key`.
 * The message is the raw signing preimage (HA-16), not pre-hashed.
 */
export function ed25519Verify(
  message: Buffer,
  sig64: Buffer,
  rawKey32: Buffer,
): boolean {
  try {
    const keyObj = createPublicKey({
      key: Buffer.concat([SPKI_PREFIX, rawKey32]),
      format: "der",
      type: "spki",
    });
    return nodeVerify(null, message, keyObj, sig64);
  } catch {
    return false;
  }
}
